security/application.md

```diff
@@ -0,0 +1,555 @@
+    > by applying a set of common transformations such as increasing a number
+    > in the password. This practice provides a false sense of security if any
+    > of the previous secrets has been compromised since attackers can apply
+    > these same common transformations.
+- Allow passwords to be more complex. Whether you want to enforce light
+  password complexity rules or not ([NIST discourages password complexity
+  rules](https://pages.nist.gov/800-63-FAQ/#q-b6)), allow for passwords longer
+  or otherwise more complex than you expect. When possible, treat it as bytes
+  that are immediately hashed and stored.
+
+A good password has a couple of properties:
+
+- Complex enough to be hard to crack through a boring enumeration attack.
+- Able to be changed when compromised.
+- Unique to the account.
+- Can be stored securely by the user (in their own head, in a password manager,
+  in a locked notebook, etc.).
+- Can be kept as a secret.
+- Leaking the password only threatens the security of that password.
+
+Biometrics (iris scan, face recognition, thumbprint reader, etc.) violate most
+of those qualities. Biometrics are useful for identity but incorrect to use as
+an authentication secret.
+
+### Timing attacks
+
+An attacker can learn a lot from _how long_ it takes to be denied access. If
+it's instant, that means the input didn't even pass validation; if it's kinda
+long, that means that the input got past validation and computed a hash but one
+of the first few characters of the hash were incorrect; a longer delay means
+that most of the hash was right. Knowing how much of the hash was right allows
+the attacker to narrow the attack space.
+
+The solution: use a constant-time equality check for comparing the hashed
+values. Bcrypt libraries ship with a function that does everything for you.
+ActiveSupport ships with
+[`secure_compare`](https://api.rubyonrails.org/classes/ActiveSupport/SecurityUtils.html#method-c-secure_compare)
+for constant-time comparisons.
+Worst case: pad the string to a fixed length then make sure your loop goes
+through every character even after you know the answer.
+
+## Multi-factor authentication (2FA)
+
+Given an email and password, you can authenticate as a user any time you wish.
+If someone were to mistakenly use the same password for multiple services, it
+is as strong as the least secure of those services: if the password were
+leaked, the password for all of those accounts are leaked at once.
+
+We can mitigate these kinds of attacks by requiring a second security factor --
+for example, a second password. We can go further by defining passwords of
+different nature:
+
+- Something you know, such as a string of letters.
+- Something you are, such as biometrics.
+- Something you have, such as a phone.
+
+We can use a HOTP or TOTP algorithm to send and verify short codes out of band
+to something the user has, such as via email, SMS, or an external program.
```

Comment:
There have been recent suggestions against SMS 2FA; it might be worth acknowledging them here. At the same time, SMS 2FA is better than no 2FA at all.

Example: https://www.makeuseof.com/tag/two-factor-authentication-sms-apps/